### PR TITLE
[sort] index build streams sorted edits

### PR DESCRIPTION
**Pull request by max-hoffman**
Use sorting to skip more steps building a prolly map. Shaves maybe 20-25% off of external index rebuilds.

This also fixes a bug where we were incorrectly using only the prefix descriptor to sort secondary index keys.

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 99.999832</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>edd9577</TD>
<TD>not ok</TD>
<TD>10</TD>
</TR>
<TR><TD>edd9577</TD>
<TD>ok</TD>
<TD>5937447</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>edd9577</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.999832</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 99.982888</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>1c0ee32</TD>
<TD>not ok</TD>
<TD>1016</TD>
</TR>
<TR><TD>1c0ee32</TD>
<TD>ok</TD>
<TD>5936441</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>1c0ee32</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.982888</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>886efa1</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>886efa1</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/table/editor/creation/external_build_index.go [hunk 1]
```diff
@@ -103,40 +104,50 @@ func BuildProllyIndexExternal(
...
-		key, err := it.Next(ctx)
+		curKey, err := t.iter.Next(ctx)
 		if err != nil {
-			if errors.Is(err, io.EOF) {
-				break
+			if !errors.Is(err, io.EOF) {
+				t.err = err
 			}
-			return nil, err
+			return nil, nil
```
Comment (reltuk) [inline go/libraries/doltcore/table/editor/creation/external_build_index.go:140]: Why isn't the caller checking `tupIter.err` if this is our error reporting mechanism here?

**Review by reltuk — APPROVED** [review]
LGTM.

As mentioned in the PR, I think we missed a check at callsite on `tupIter.err`

File: go/libraries/doltcore/table/editor/creation/external_build_index.go (hunk 1 above)
Comment (max-hoffman) [reply go/libraries/doltcore/table/editor/creation/external_build_index.go:143]: tru

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>f8d1c43</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>f8d1c43</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>1a9ed83</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>1a9ed83</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Pull request merged**